asap/rough_align/fit_multiple_solves.py

```diff
@@ -0,0 +1,367 @@
+
+def baseline_vertices(xmin, xmax, ymin, ymax, nx, ny):
+    xt, yt = np.meshgrid(
+            np.linspace(xmin, xmax, nx),
+            np.linspace(ymin, ymax, ny))
+    return np.vstack((xt.flatten(), yt.flatten())).transpose()
+
+
+def tps_from_vertices(vertices):
+    tf = renderapi.transform.ThinPlateSplineTransform()
+    tf.ndims = 2
+    tf.nLm = vertices.shape[0]
+    tf.aMtx = np.array([[0.0, 0.0], [0.0, 0.0]])
+    tf.bVec = np.array([0.0, 0.0])
+    tf.srcPts = vertices.transpose()
+    tf.dMtxDat = np.zeros_like(tf.srcPts)
+    return tf
+
+
+def append_tps_tform(tspec, npts=5, ext=0.05):
+    bb = tspec.bbox_transformed()
+    xmin, ymin = bb.min(axis=0)
+    xmax, ymax = bb.max(axis=0)
+    w, h = bb.ptp(axis=0)
+    vert = baseline_vertices(
+        xmin - ext * w,
+        xmax + ext * w,
+        ymin - ext * h,
+        ymax + ext * h,
+        npts, npts
+    )
+    tform = tps_from_vertices(vert)
+    tspec.tforms.append(tform)
+
+def thin_plate_spline_solve(aff_result_rts,new_matches):
+    
+    tpsadded_aff_result_rts = copy.deepcopy(aff_result_rts)
+    for tspec in tpsadded_aff_result_rts.tilespecs:
+        append_tps_tform(tspec, npts=5)
+    
+    tps_transform_name = "ThinPlateSplineTransform"
+    tps_order = 2
+    tps_pair_depth = 4
+    tps_fullsize = False
+    tps_transform_apply = [0]
+    tps_regularization_dict = {
+        "translation_factor": 1e-10,
+        "default_lambda": 1e7,
+        "freeze_first_tile": False,
+        "thinplate_factor": 5e7
+    }
+    tps_matrix_assembly_dict = {
+        "choose_random": True,
+        "depth": 3,
+        "npts_max": 500,
+        "inverse_dz": True,
+        "npts_min": 5,
+        "cross_pt_weight": 0.5,
+        "montage_pt_weight": 1.0
+    }
+
+    tps_matches = copy_matches(new_matches)  # any transform_apply modifies matches.... bummer.
+
+    tps_create_CRS_A_input = (
+        tpsadded_aff_result_rts, tps_matches, tps_transform_name,
+        tps_transform_apply, tps_regularization_dict, tps_matrix_assembly_dict, tps_order, tps_fullsize)
+    return tps_create_CRS_A_input,tps_regularization_dict
+
+
+class FitMultipleSolves(argschema.ArgSchemaParser):
+    default_schema = FitMultipleSolvesSchema
+    default_output_schema = FitMultipleSolvesOutputSchema
+
+    def apply_transform(self,new_matches,rts,app):
+        app_solver = {'rigid': rigid_solve,
+            'translate': translation_solve,
+            'affine': affine_solve,
+            'tps': thin_plate_spline_solve}
+        app_input,app_reg_dict = app_solver[app](rts,new_matches)
+        app_fr, app_draft_resolvedtiles = bigfeta.bigfeta.create_CSR_A_fromobjects(*app_input, return_draft_resolvedtiles=True)
+        app_result_rts = copy.deepcopy(app_draft_resolvedtiles)
+        app_reg = scipy.sparse.diags(
+            [numpy.concatenate([ts.tforms[-1].regularization(app_reg_dict) for ts in app_draft_resolvedtiles.tilespecs])],
```

Comment:
```suggestion
            [np.concatenate([ts.tforms[-1].regularization(app_reg_dict) for ts in app_draft_resolvedtiles.tilespecs])],
```
numpy is only imported as np in this import block